presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeJoinsOverMapColumns.java

```diff
@@ -0,0 +1,490 @@
+        {
+            return rewriteScanFilterProject(scanNode, context);
+        }
+
+        private PlanNode rewriteScanFilterProject(PlanNode planNode, RewriteContext<JoinContext> context)
+        {
+            Set<VariableReferenceExpression> joinKeys = context.get().getJoinKeys();
+
+            if (joinKeys.size() == 0 || context.get().getNumJoins() < 2) {
+                return planNode;
+            }
+
+            if (!ImmutableSet.copyOf(planNode.getOutputVariables()).containsAll(joinKeys)) {
+                // not all join keys are in the plan node: abort rewrite
+                return planNode;
+            }
+
+            List<VariableReferenceExpression> groupingKeys = joinKeys.stream().collect(Collectors.toList());
+
+            AggregationNode agg = new AggregationNode(
+                    planNode.getSourceLocation(),
+                    planNodeIdAllocator.getNextId(),
+                    planNode,
+                    ImmutableMap.of(),
+                    singleGroupingSet(groupingKeys),
+                    ImmutableList.of(),
+                    SINGLE,
+                    Optional.empty(),
+                    Optional.empty());
+
+            Map<VariableReferenceExpression, VariableReferenceExpression> varMap = new HashMap<>();
+            ImmutableList.Builder<JoinNode.EquiJoinClause> equiJoinClausesBuilder = ImmutableList.builder();
+
+            for (VariableReferenceExpression var : joinKeys) {
+                VariableReferenceExpression newVar = planVariableAllocator.newVariable(var.getName(), var.getType());
+                varMap.put(var, newVar);
+                equiJoinClausesBuilder.add(new JoinNode.EquiJoinClause(newVar, var));
+            }
+            PlanNode newPlanNode = clonePlanNode(planNode, joinKeys, varMap, planNodeIdAllocator);
+            context.get().setFilterScanNode(newPlanNode);
+            context.get().addOuterJoinClauses(equiJoinClausesBuilder.build());
+            return agg;
+        }
+
+        private PlanNode clonePlanNode(PlanNode planNode, Set<VariableReferenceExpression> joinKeys, Map<VariableReferenceExpression, VariableReferenceExpression> varMap, PlanNodeIdAllocator idAllocator)
+        {
+            if (planNode instanceof TableScanNode) {
+                TableScanNode scanNode = (TableScanNode) planNode;
+                return cloneTableScan(joinKeys, varMap, scanNode);
+            }
+            else if (planNode instanceof FilterNode) {
+                return cloneFilterNode((FilterNode) planNode, joinKeys, varMap, idAllocator);
+            }
+            return cloneProjectNode((ProjectNode) planNode, joinKeys, varMap, idAllocator);
+        }
+
+        private FilterNode cloneFilterNode(FilterNode filterNode, Set<VariableReferenceExpression> joinKeys, Map<VariableReferenceExpression, VariableReferenceExpression> varMap, PlanNodeIdAllocator idAllocator)
```

Comment:
Let's move these clone* methods to PlannerUtils so we can use it for other cases as well. This could eventually evlove into cloneSubplan